helpers/external_acl/session/ext_session_acl.8
.if !'po4a'hide' .TH ext_session_acl 8 "9 October 2011"
.
.SH NAME
ext_session_acl \- Squid session tracking external acl helper.
.PP
Version 1.2
.
.SH SYNOPSIS
.if !'po4a'hide' .B ext_session_acl
.if !'po4a'hide' .B "[\-t"
timeout
.if !'po4a'hide' .B "] [\-b"
database
.if !'po4a'hide' .B "] [\-a]"
.
.SH DESCRIPTION
.B ext_session_acl
maintains a concept of sessions by monitoring requests
and timing out sessions. The timeout is based either on idle use (
.B \-t
) or a fixed period of time (
.B \-T
). The former is suitable for displaying terms and conditions to a user; the
latter is suitable for the display of advertisements or other notices (both as a
splash page \- see config examples in the wiki online). The session helper can also be used
to force users to re\-authenticate if the 
.B %LOGIN 
and 
.B \-a
are both used.
.
.SH OPTIONS
.if !'po4a'hide' .TP 12
.if !'po4a'hide' .B "\-t timeout"
Idle timeout for any session. The default if not specified (set to 3600 seconds).
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-T timeout"
Fixed timeout for any session. This will end the session after the timeout regardless
of a user's activity. If used with
.B active
mode, this will terminate the user's session after
.B timeout
, after which another
.B LOGIN
will be required.
.B LOGOUT
will reset the session and timeout.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-b path"
.B Path
to persistent database. If a file is specified then that single file is
used as the database. If a path is specified, a Berkeley DB database
environment is created within the directory. The advantage of the latter
is better database support between multiple instances of the session
helper. Using multiple instances of the session helper with a single
database file will cause synchronization problems between processes.
If this option is not specified the session details will be kept in
memory only and all sessions will reset each time Squid restarts its
helpers (Squid restart or rotation of logs).
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B \-a
Active mode. In this mode sessions are started by evaluating an
acl with the argument
.B LOGIN
, or terminated by the argument
.B LOGOUT \.
Without this flag the helper automatically starts the session after
the first request.
.SH CONFIGURATION
.PP
The
.B ext_session_acl
helper is a concurrent helper; therefore, the concurrency= option
.B must
be specified in the configuration.
.PP
Passive session configuration example using the default automatic mode
.if !'po4a'hide' .RS
.if !'po4a'hide' .B external_acl_type session ttl=300 negative_ttl=0 children=1 concurrency=200 %LOGIN /usr/local/squid/libexec/ext_session_acl
.if !'po4a'hide' .br
.if !'po4a'hide' .B acl session external session
.if !'po4a'hide' .br
.if !'po4a'hide' .B http_access deny !session
.if !'po4a'hide' .br
.if !'po4a'hide' .B deny_info http://your.server.example.com/bannerpage?url=%s session
.if !'po4a'hide' .RE
.PP
Then set up 
.B http://your.server.example.com/bannerpage 
to display a session startup page and then redirect the user back to the requested URL given in the url query parameter.
.
.SH AUTHOR
This program and documentation was written by
.if !'po4a'hide' .I Henrik Nordstrom <henrik@henriknordstrom.net>
.if !'po4a'hide' .I Andrew Beverley <andy@andybev.com>
.
.SH COPYRIGHT
.PP
 * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
 *
 * Squid software is distributed under GPLv2+ license and includes
 * contributions from numerous individuals and organizations.
 * Please see the COPYING and CONTRIBUTORS files for details.
.PP
This program and documentation is copyright to the authors named above.
.PP
Distributed under the GNU General Public License (GNU GPL) version 2 or later (GPLv2+).
.
.SH QUESTIONS
Questions on the usage of this program can be sent to the
.I Squid Users mailing list
.if !'po4a'hide' <squid-users@squid-cache.org>
.
.SH REPORTING BUGS
Bug reports need to be made in English.
See http://wiki.squid-cache.org/SquidFaq/BugReporting for details of what you need to include with your bug report.
.PP
Report bugs or bug fixes using http://bugs.squid-cache.org/
.PP
Report serious security bugs to
.I Squid Bugs <squid-bugs@squid-cache.org>
.PP
Report ideas for new improvements to the
.I Squid Developers mailing list
.if !'po4a'hide' <squid-dev@squid-cache.org>
.
.SH SEE ALSO
.if !'po4a'hide' .BR squid "(8), "
.if !'po4a'hide' .BR GPL "(7), "
.br
The Squid FAQ wiki
.if !'po4a'hide' http://wiki.squid-cache.org/SquidFaq
.br
The Squid Configuration Manual
.if !'po4a'hide' http://www.squid-cache.org/Doc/config/


ChangeLog
Changes to squid-4.0.23 (19 Jan 2018):

	- Bug 4715: security_file_certgen: Remove -g and -n options docs
	- Bug 4679: User names not sent to url_rewrite_program
	- Bug 4631: security_file_certgen helper without disk cache
	- Bug 3911: clang -fsanitize warnings
	- Bug 2378: Duplicates in selected peer destinations
	- Nettle v3.4 support
	- Fix Squid FTP server dying because of an unhandled exception
	- Automatically revive hopeless kids on reconfigure and after a timeout
	- Fix %<Hs, %<pt, %<tt, %<bs calculation bugs for error responses
	- ... and many documentation updates
	- ... and some stability fixes

Changes to squid-4.0.22 (07 Dec 2017):

	- Regression fix: Relay peer CONNECT error status line and headers to clients
	- Bug 4767: SMP breaks IPv6 SNMP and cache manager queries
	- Bug 4718: support filling raw buffer space of shared SBufs
	- Bug 4648: object revalidation for HTTPS scheme
	- Bug 4616: store_client.cc:92: "mem" assertion
	- Bug 2821: ignore Content-Range in non-206 responses
	- HTCP: Ignore packets with invalid URI
	- TLS: Validate the shortest certificate chain
	- TLS: Add checks for OpenSSL 1.1.0f API changes
	- TLS: Fix reporting of validation errors for downloaded intermediate certs
	- TLS: Fix SSL certificate cache refresh and collision handling
	- Fix backwards compatibility for Squid-3.5 external_acl_type formats
	- Fix invalid mime icon URLs in cache
	- Do not die silently when dying early
	- Docs: update translation files

Changes to squid-4.0.21 (02 Jul 2017):

	- Bug 4730: segfault while processing internal HTTP requests
	- Bug 4492: Chunk extension parser is too pedantic
	- Bug 1961: Redesign urlParse() API
	- TLS: recognise tls:: namespace on logformat tokens
	- SSL-Bump: tproxy does not spoof spliced connections
	- security_file_certgen: collapse queued requests
	- Add a basic apparmour profile
	- Add transaction_initiator ACL for detecting various unusual transactions
	- Add ssl::server_name options to control matching logic
	- Support for --long-acl-options
	- Do not die silently when dying via std::terminate()
	- Fix option --foreground to implement expected behavior
	- Translations: update .po and .pot to latest texts
	- ... and some documentation updates
	- ... and many code cleanup and stability fixes
	- ... and all fixes from 3.5.27

Changes to squid-4.0.20 (01 Jun 2017):

	- Bug 4692: SslBump breaks intercepted IPv6 connections
	- Bug 4682: ignoring http_access deny when client-first bumping mode is used
	- Bug 4662: build errors with LibreSSL 2.4.4
	- Bug 4659: sslproxy_foreign_intermediate_certs does not work
	- Bug 4321: ssl_bump terminate does not terminate at step1
	- Add 'has' ACL
	- Do not forward HTTP requests to dead idle peers
	- Do not unconditionally revive dead peers after a DNS refresh
	- Make PID file check/creation atomic to avoid associated race conditions
	- Count failures and use peer-specific connect timeouts when tunneling
	- SSL-Bump: Fix crashes when server-first bumping mode is used with openSSL-1.1.0
	- eCAP: Fix empty header handling in Ecap::HeaderRep::hasAny()
	- SSL-Bump: Second adaptation missing for CONNECTs
	- ext_session_acl: cope with new logformat inputs
	- ... and some documentation updates
	- ... and some code stability fixes
	- ... and all fixes from 3.5.26

Changes to squid-4.0.19 (02 Apr 2017):

	- Bug 4674: delay_parameters for class 3 and 4 assertion failed
	- Bug 4671: GCC 7 compile errors
	- Bug 4663: GCC 5+ compile errors with optimization level -O3
	- Bug 4657: delay IDENT until after PROXY protocol handling
	- Bug 4610: cleanup of BerkleyDB related checks
	- squidclient: Fix missing error handling on PUT
	- digest_ldap_auth: Add -r option to clamp the realm to a fixed value
	- TLS: initial GnuTLS support for encrypted server connections
	- Fix appending Http::HdrType::VIA code
	- Fix URI scheme case-sensitivity treatment
	- Fix two read-ahead problems related to delay pools (or lack thereof)
	- Detail swapfile header inconsistencies
	- ... and several build fixes
	- ... and many code polishing updates
	- ... and all fixes from 3.5.25

Changes to squid-4.0.18 (06 Feb 2017):

	- Bug 4661: compile error 'warning: _XPG4_2 redefined' with GCC on Solaris 10
	- Bug 4636: assertion 'byteCount > 0 && byteCount <= inBuf.length()'
	- Bug 4610 partial: compile errors on Solaris 11.3 with Oracle Studio 12.5
	- Bug 4599: support OpenSSL 1.1
	- squidclient: link GnuTLS library debugs to -v level display
	- Fix GCC6: unused local variable 'weInitiatedThisClosure'
	- ... and some code polishing
	- ... and some copyright updates
	- ... and all fixes from 3.5.24

Changes to squid-4.0.17 (16 Dec 2016):

	- Bug 4630: user credentials cache cleanup not re-scheduled
	- Bug 4610 partial: compile errors on Solaris 11.3 with Oracle Studio 12.5
	- Bug 4599 partial: initial support for OpenSSL v1.1
	- TLS: Support tunneling of bumped non-HTTP traffic
	- ... and many code polishing and performance updates
	- ... and some documentation updates
	- ... and some fixes from 3.5.23

Changes to squid-4.0.16 (30 Oct 2016):

	- Avoid segfaults when lacking the server name for certificate validator
	- HTTP: initial support for Cache-Control:immutable
	- Fix ssl::server_name ACL
	- ... and many code polishing updates
	- ... and some fixes from 3.5.23

Changes to squid-4.0.15 (09 Oct 2016):

	- Regression fix crash on reconfigure with TOS/DiffServ/MARK configured
	- Bug 4610: compile errors on Solaris 11.3 with Oracle Studio 12.5
	- Bug 4581: Secure ICAP segfault in checkForMissingCertificates
	- Bug 4578: changes required to install squid.service
	- Fix crash on shutdown while cleaning up idle ICAP connections
	- Fix memory leak of Downloader-related objects
	- HTTP/1.1: handle syntactically valid requests with unsupported HTTP versions
	- Log TCP client port for error:transaction-end-before-headers and such
	- ... and many portability and build fixes
	- ... and some documentation updates
	- ... and all fixes from 3.5.22


helpers/external_acl/session/ext_session_acl.cc
/*
 * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
 *
 * Squid software is distributed under GPLv2+ license and includes
 * contributions from numerous individuals and organizations.
 * Please see the COPYING and CONTRIBUTORS files for details.
 */

/*
 * ext_session_acl: Squid external acl helper for tracking sessions
 *
 * Copyright (C) 2006 Henrik Nordstrom <henrik@henriknordstrom.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.
 */

#if HAVE_CONFIG_H
#include "squid.h"
#endif
#include "helpers/defines.h"

#include <cstdlib>
#include <cstring>
#include <ctime>
#if HAVE_DB_H
#include <db.h>
#endif
#include <fcntl.h>
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <sys/types.h>
#include <sys/stat.h>
#if HAVE_UNISTD_H
#include <unistd.h>
#endif

/* At this point all Bit Types are already defined, so we must
   protect from multiple type definition on platform where
   __BIT_TYPES_DEFINED__ is not defined.
 */
#ifndef        __BIT_TYPES_DEFINED__
#define        __BIT_TYPES_DEFINED__
#endif

static int session_ttl = 3600;
static int fixed_timeout = 0;
char *db_path = NULL;
const char *program_name;

DB *db = NULL;
DB_ENV *db_env = NULL;

static void init_db(void)
{
    struct stat st_buf;

    if (db_path) {
        if (!stat(db_path, &st_buf)) {
            if (S_ISDIR (st_buf.st_mode)) {
                /* If directory then open database environment. This prevents sync problems
                    between different processes. Otherwise fallback to single file */
                db_env_create(&db_env, 0);
                if (db_env->open(db_env, db_path, DB_CREATE | DB_INIT_MPOOL | DB_INIT_LOCK , 0666)) {
                    fprintf(stderr, "FATAL: %s: Failed to open database environment in '%s'\n", program_name, db_path);
                    db_env->close(db_env, 0);
                    exit(1);
                }
                db_create(&db, db_env, 0);
            }
        }
    }

    if (db_env) {
        if (db->open(db, NULL, "session", NULL, DB_BTREE, DB_CREATE, 0666)) {
            fprintf(stderr, "FATAL: %s: Failed to open db file '%s' in dir '%s'\n",
                    program_name, "session", db_path);
            db_env->close(db_env, 0);
            exit(1);
        }
    } else {
        db_create(&db, NULL, 0);
        if (db->open(db, NULL, db_path, NULL, DB_BTREE, DB_CREATE, 0666)) {
            fprintf(stderr, "FATAL: %s: Failed to open session db '%s'\n", program_name, db_path);
            exit(1);
        }
    }
}

static void shutdown_db(void)
{
    db->close(db, 0);
    if (db_env) {
        db_env->close(db_env, 0);
    }
}

int session_is_active = 0;

static int session_active(const char *details, size_t len)
{
    DBT key = {0};
    DBT data = {0};
    key.data = (void *)details;
    key.size = len;
    if (db->get(db, NULL, &key, &data, 0) == 0) {
        time_t timestamp;
        if (data.size != sizeof(timestamp)) {
            fprintf(stderr, "ERROR: %s: CORRUPTED DATABASE (%s)\n", program_name, details);
            db->del(db, NULL, &key, 0);
            return 0;
        }
        memcpy(&timestamp, data.data, sizeof(timestamp));
        if (timestamp + session_ttl >= time(NULL))
            return 1;
    }
    return 0;
}

static void session_login(const char *details, size_t len)
{
    DBT key = {0};
    DBT data = {0};
    key.data = (void *)details;
    key.size = len;
    time_t now = time(NULL);
    data.data = &now;
    data.size = sizeof(now);
    db->put(db, NULL, &key, &data, 0);
}

static void session_logout(const char *details, size_t len)
{
    DBT key = {0};
    key.data = (void *)details;
    key.size = len;
    db->del(db, NULL, &key, 0);
}

static void usage(void)
{
    fprintf(stderr, "Usage: %s [-t|-T session_timeout] [-b dbpath] [-a]\n", program_name);
    fprintf(stderr, "	-t sessiontimeout	Idle timeout after which sessions will be forgotten (user activity will reset)\n");
    fprintf(stderr, "	-T sessiontimeout	Fixed timeout after which sessions will be forgotten (regardless of user activity)\n");
    fprintf(stderr, "	-b dbpath		Path where persistent session database will be kept\n");
    fprintf(stderr, "	-a			Active mode requiring LOGIN argument to start a session\n");
}
int main(int argc, char **argv)
{
    char request[HELPER_INPUT_BUFFER];
    int opt;
    int default_action = 1;

    program_name = argv[0];

    while ((opt = getopt(argc, argv, "t:T:b:a?")) != -1) {
        switch (opt) {
        case 'T':
            fixed_timeout = 1;
        case 't':
            session_ttl = strtol(optarg, NULL, 0);
            break;
        case 'b':
            db_path = optarg;
            break;
        case 'a':
            default_action = 0;
            break;
        case '?':
            usage();
            exit(0);
            break;
        }
    }

    setbuf(stdout, NULL);

    init_db();

    while (fgets(request, HELPER_INPUT_BUFFER, stdin)) {
        int action = 0;
        const char *channel_id = strtok(request, " ");
        char *detail = strtok(NULL, "\n");
        if (detail == NULL) {
            // Only 1 paramater supplied. We are expecting at least 2 (including the channel ID)
            fprintf(stderr, "FATAL: %s is concurrent and requires the concurrency option to be specified.\n", program_name);
            shutdown_db();
            exit(1);
        }
        char *lastdetail = strrchr(detail, ' ');
        size_t detail_len = strlen(detail);
        if (lastdetail) {
            if (strcmp(lastdetail, " LOGIN") == 0) {
                action = 1;
                detail_len = (size_t)(lastdetail-detail);
                *lastdetail = '\0';
            } else if (strcmp(lastdetail, " LOGOUT") == 0) {
                action = -1;
                detail_len = (size_t)(lastdetail-detail);
                *lastdetail = '\0';
            }
        }
        if (action == -1) {
            session_logout(detail, detail_len);
            printf("%s OK message=\"Bye\"\n", channel_id);
        } else if (action == 1) {
            session_login(detail, detail_len);
            printf("%s OK message=\"Welcome\"\n", channel_id);
        } else if (session_active(detail, detail_len)) {
            if (fixed_timeout == 0) {
                session_login(detail, detail_len);
            }
            printf("%s OK\n", channel_id);
        } else if (default_action == 1) {
            session_login(detail, detail_len);
            printf("%s ERR message=\"Welcome\"\n", channel_id);
        } else {
            printf("%s ERR message=\"No session available\"\n", channel_id);
        }
    }
    shutdown_db();
    return 0;
}

